pkg/diagnose/check.go

```diff
@@ -24,10 +26,92 @@ func init() {
 	diagnosis.Register("check-datadog", diagnose)
 }
 
+func diagnose(diagCfg diagnosis.Config) []diagnosis.Diagnosis {
+	if diagCfg.RunningInAgentProcess {
+		return diagnoseInAgentProcess(diagCfg)
+	}
+
+	return diagnoseInCLIProcess(diagCfg)
+}
+
+func partitionChecks(checks []check.Check) [][]check.Check {
+	instancesMap := map[string][]check.Check{}
+
+	// partition per check (instances of a check groupped by check/integration)
+	for _, c := range checks {
+		checkName := c.String()
+		instances, exists := instancesMap[checkName]
+		if exists {
+			instances = append(instances, c)
+		} else {
+			instances = make([]check.Check, 1)
+			instances[0] = c
+		}
+		instancesMap[checkName] = instances
+	}
+
+	// copy map to array
+	instancesArr := make([][]check.Check, 0, len(instancesMap))
+	for _, instances := range instancesMap {
+		instancesArr = append(instancesArr, instances)
+	}
+
+	// sort this array
+	sort.Slice(instancesArr, func(i, j int) bool {
+		return instancesArr[i][0].String() < instancesArr[j][0].String()
+	})
+
+	return instancesArr
+}
+
+func getInstanceDiagnoses(instance check.Check) []diagnosis.Diagnosis {
+
+	// Get diagnoses
+	diagnoses, err := instance.GetDiagnoses()
+	if err == nil && len(diagnoses) > 0 {
+		// Set category as check name if it was not set
+		for i, d := range diagnoses {
+			if len(d.Category) == 0 {
+				diagnoses[i].Category = instance.String()
+			}
+		}
+	} else if err != nil {
+		// Check Diagnose method return error
+		diagnoses = append(diagnoses, diagnosis.Diagnosis{
+			Result:    diagnosis.DiagnosisUnexpectedError,
+			Name:      string(instance.ID()),
+			Category:  instance.String(),
+			Diagnosis: "Check Dianose failes with unexpected errors",
+			RawError:  err,
+		})
+	}
+
+	return diagnoses
+}
+
+func diagnoseInAgentProcess(diagCfg diagnosis.Config) []diagnosis.Diagnosis {
```

Comment:
`dialCfg` appears to be unused. Can it be removed?

Why second way of running diagnostics at all? Would it make sense to merge this code with the other one (that seems to honor the configuration).

Reply:
Eventually diagCfg may be passed further down the line to the diagnose check code. Removed it per your suggestion and will add it when corresponding code will appear, thank you.

Regarding running in a different existing function, there are a few issues/ First, diagnose code may be different (e.g. getting checks from Agent vs non-Agent process is different). Second, I am using the same API Run() and if will have identical method then I will get into recursion. There are probably better ways to handling that and I will look for improvements in future.